### PR TITLE
Deliverable 5 david

### PR DESCRIPTION
**Pull request by 7thStringofZhef**
This isn't ready for merger, so don't accept it, just review what's here. At a bare minimum, I need to hear back from Yiran about two parameters. It's also a pain to fully test with all of the hashing, the required Redis server, etc, so I've gotta do that as best I can when I find time. This is not going to be fully integrated with the frontend, we've gone too far off the rails for that. I'm just trying to provide as much functional value as I can that the customer can put in a more fully-realized solution.

### CODE REVIEW

**Comment by 7thStringofZhef** [comment]
Alright, I'm about as happy with it as I'm going to be. Merge this once you finish reviewing it.

File: README.md [hunk 1]
```diff
@@ -1,39 +1,98 @@
...
-1. Sessions are enabled. You can login to a session using any username and password combination, and it will remember your session when you refresh (doesn't look like much when running, though, since not integrated with frontend)
-2. Email is enabled. 
-3. Created a method of receiving full files for jobs instead of just filenames. Still not integrated with frontend
-
-#Outstanding issues in deliverable 4
-Unfortunately, I have not been able to solve the test issues from Deliverable 3 and am working with our QA to solve them.
-1. Unit tests fail for creating a User where one already exists(expected it to throw exception) and for updating a nonexistent user (expected exception)
-2. Currently reworking login and logout functionality, so although login and logout controllers still function, all operations can be performed without logging in
-3. We need integration to see the effects of sessions and full files.
+#Fully-implmented features
```
Comment (AdityaMahadevan) [inline README.md]: This README looks perfect, exactly what they said they were looking for in documentation.

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/API/OSGTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		
+		HttpClient httpClient = HttpClientBuilder.create().build();
+		
+		try {
+			
+			HttpPost request = new HttpPost("http://localhost:8080/freesurfer/job");
+			StringEntity parameters = new StringEntity(JSONString);
+			request.addHeader("content-type", "application/x-www-form-urlencoded");
+			request.setEntity(parameters);
+			HttpResponse response = httpClient.execute(request);
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/API/OSGTest.java]: :+1:

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/API/OSGTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			return IOUtils.toString(responseStream, "UTF-8");
+		}
+		catch (Exception e) {
+			System.out.println("Exception: " + e);
+			return "";
+		}
+	}
+	
+	/*
+	public static void main(String[] args) {
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/API/OSGTest.java]: Might want to get rid of this part completely or comment why its still here.

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/config/SecurityConfig.java [hunk 4]
```diff
@@ -1,21 +1,60 @@
...
 	@Autowired
+	OSGAuthenticationProvider authenticationProvider; //Inject an instance of our custom authentication
+	
+	@Autowired
+	OSGUserDetailsService userDetailsService; //Inject an instance of custom user details service
+	
+	//Describes how security should work
+	//Note that because sessions are enabled, if I receieve a valid token, then I skip other authentication
+	@Override
+	protected void configure(HttpSecurity http) throws Exception {
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/config/SecurityConfig.java]: :+1: for comments, easy to read code

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/config/SimpleCORSFilter.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+		response.setHeader("Access-Control-Allow-Origin", "*");
+		response.setHeader("Access-Control-Allow-Methods", "POST, GET, PUT, OPTIONS, DELETE");
+		response.setHeader("Access-Control-Max-Age", "3600");
+		response.setHeader("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+		chain.doFilter(req, res);
+	}
+
+	public void init(FilterConfig filterConfig) {}
+
+	public void destroy() {}
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/config/SimpleCORSFilter.java]: init and destroy don't do anything?

Comment (7thStringofZhef) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/config/SimpleCORSFilter.java]: They're just necessary functions of filter. If left alone, Spring injects the default filter functionality

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/controllers/job/JobController.java [hunk 6]
```diff
@@ -19,28 +21,55 @@
...
+		{
+			try{
+				info = mapper.writeValueAsString(job);
+			}
+			catch(Exception e){e.printStackTrace(); return null;}
+			return new ResponseEntity<String>(info, HttpStatus.OK); //Return user's info as sub-JSON and "Ok"
+		}
 	}
 	
+	/*
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/controllers/job/JobController.java]: Again, may want to either get rid of this or comment why the code is commented out.

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/controllers/job/JobController.java [hunk 7]
```diff
@@ -48,36 +77,79 @@
...
+			return new ResponseEntity<String>(info, HttpStatus.INTERNAL_SERVER_ERROR);
+		}
+		//Need to find integer user_id from username
+		long userId = userService.findByUsername(principal.getName()).getid();
+		String hashedPwd = userService.findByUsername(principal.getName()).getpassword();
+		
+		if(!file.isEmpty())
+		{
+			try{
+				OSGTest.sendJob((int)userId, hashedPwd, filename, !multicore, Long.toString(savedJob.getid())); //Empty string is token
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/controllers/job/JobController.java]: :+1:

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/controllers/user/UserController.java [hunk 8]
```diff
@@ -19,91 +29,178 @@
 import cs1530.osgfreesurfer.routes.User;
 import cs1530.osgfreesurfer.services.UserService;
 
+//User url functions
+//Note that in all cases where it takes a token, the token is already implicitly handled by security
+//Token allows it to choose the specific session that is logged in and thus find the correct user
 @RestController
 public class UserController
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/controllers/user/UserController.java]: :+1: :100:

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/security/CustomPasswordEncoder.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+import org.bouncycastle.util.encoders.Hex;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+public class CustomPasswordEncoder{
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/security/CustomPasswordEncoder.java]: :+1:

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/security/OSGAuthenticationProvider.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+import cs1530.osgfreesurfer.services.UserService;
+
+//Configure my authentication provider
+//The way that this particular one works is that:
+//AuthenticationManager calls provider.authenticate(userDetails, token)
+//OSGAuthenticationProvider calls retrieveUser, then additionalAuthenticationChecks
+//Returns whether or not user was authenticated
+@Component
+public class OSGAuthenticationProvider
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/security/OSGAuthenticationProvider.java]: :+1:

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/security/OSGUserDetailsService.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            // Not found...
+            throw new UsernameNotFoundException(
+                    "User " + username + " not found.");
+        }
+
+        Collection<GrantedAuthority> grantedAuthorities = new ArrayList<GrantedAuthority>();
+        grantedAuthorities.add(new SimpleGrantedAuthority("ROLE_USER")); //All people have user authority
+
+        //Messy, but I call my users Users as well, and what I really need is Spring's definition
+        //Explaining booleans:
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/security/OSGUserDetailsService.java]: :100:

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/services/UserServiceBean.java [hunk 12]
```diff
@@ -36,6 +37,12 @@ public User findOne(long id) {
 		User user = userRepository.findOne(id);
 		return user;
 	}
+	
+	public User findByUsername(String username)
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/services/UserServiceBean.java]: :+1:

File: skeleton-web-services/src/main/java/cs1530/osgfreesurfer/services/UserServiceBean.java [hunk 13]
```diff
@@ -62,43 +69,21 @@ public User update(User user) {
...
 		SimpleMailMessage msg = new SimpleMailMessage();
 		msg.setTo(email);
 		msg.setSubject("Email test");
 		msg.setText("Hello, this is a test of spring email system");
 		javaMailService.send(msg);
 	}
+	
+	//Used to asynchronously update job status. Gets a JSON. In future iterations, schedule this as a recurrent process and use the info to update jobs in the database
+	@Override
+	public String getJobsStatus(long id)
```
Comment (AdityaMahadevan) [inline skeleton-web-services/src/main/java/cs1530/osgfreesurfer/services/UserServiceBean.java]: :+1:

**Pull request merged**